src/trio/_core/_concat_tb.py

```diff
@@ -0,0 +1,128 @@
+# modifying traceback objects, so we have to get creative...
+#
+# On CPython, we use ctypes. On PyPy, we use "transparent proxies".
+#
+# Jinja2 is a useful source of inspiration:
+#   https://github.com/pallets/jinja/blob/master/jinja2/debug.py
+
+try:
+    import tputil
+except ImportError:
+    # ctypes it is
+    import ctypes
+
+    # How to handle refcounting? I don't want to use ctypes.py_object because
+    # I don't understand or trust it, and I don't want to use
+    # ctypes.pythonapi.Py_{Inc,Dec}Ref because we might clash with user code
+    # that also tries to use them but with different types. So private _ctypes
+    # APIs it is!
+    import _ctypes
+
+    class CTraceback(ctypes.Structure):
+        _fields_: ClassVar = [
+            ("PyObject_HEAD", ctypes.c_byte * object().__sizeof__()),
+            ("tb_next", ctypes.c_void_p),
+            ("tb_frame", ctypes.c_void_p),
+            ("tb_lasti", ctypes.c_int),
+            ("tb_lineno", ctypes.c_int),
+        ]
+
+    def copy_tb(base_tb: TracebackType, tb_next: TracebackType | None) -> TracebackType:
+        # TracebackType has no public constructor, so allocate one the hard way
+        try:
+            raise ValueError
+        except ValueError as exc:
+            new_tb = exc.__traceback__
+            assert new_tb is not None
+        c_new_tb = CTraceback.from_address(id(new_tb))
+
+        # At the C level, tb_next either pointer to the next traceback or is
+        # NULL. c_void_p and the .tb_next accessor both convert NULL to None,
+        # but we shouldn't DECREF None just because we assigned to a NULL
+        # pointer! Here we know that our new traceback has only 1 frame in it,
+        # so we can assume the tb_next field is NULL.
+        assert c_new_tb.tb_next is None
+        # If tb_next is None, then we want to set c_new_tb.tb_next to NULL,
+        # which it already is, so we're done. Otherwise, we have to actually
+        # do some work:
+        if tb_next is not None:
+            _ctypes.Py_INCREF(tb_next)  # type: ignore[attr-defined]
+            c_new_tb.tb_next = id(tb_next)
+
+        assert c_new_tb.tb_frame is not None
+        _ctypes.Py_INCREF(base_tb.tb_frame)  # type: ignore[attr-defined]
+        old_tb_frame = new_tb.tb_frame
+        c_new_tb.tb_frame = id(base_tb.tb_frame)
+        _ctypes.Py_DECREF(old_tb_frame)  # type: ignore[attr-defined]
+
+        c_new_tb.tb_lasti = base_tb.tb_lasti
+        c_new_tb.tb_lineno = base_tb.tb_lineno
+
+        try:
+            return new_tb
+        finally:
+            # delete references from locals to avoid creating cycles
+            # see test_cancel_scope_exit_doesnt_create_cyclic_garbage
+            del new_tb, old_tb_frame
+
+else:
+    # http://doc.pypy.org/en/latest/objspace-proxies.html
+    def copy_tb(base_tb: TracebackType, tb_next: TracebackType | None) -> TracebackType:
+        # tputil.ProxyOperation is PyPy-only, but we run mypy on CPython
```

Comment:
Unfortunately, as I found out recently, type checkers don't have a way of saying some code is PyPy / CPython specific *anyways*... So this wouldn't work regardless. I think this comment could probably be removed (or reworded) cause it insinuates that it is possible with the current state of typing.

Reply:
reworded it